bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,401 @@
+        this.url = "http://" + ip + "/HNAP1/";
+    }
+
+    private SOAPElement moduleParameters(final String moduleID) throws SOAPException {
+        return soapFactory.createElement("ModuleID").addTextNode(moduleID);
+    }
+
+    public String queryStringState(final String action, final String name, final String module) {
+        try {
+            final Document document = post(action, newMessage(action, moduleParameters(module)), authentication);
+            return getElementByTagName(document, name);
+        } catch (CommunicationException e) {
+            logger.error("Failed to get quantity (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        } catch (SOAPException e) {
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
+        }
+        return "";
+    }
+
+    public State queryOnOffState(final String action, final String name, final String module) {
+        final String state = queryStringState(action, name, module);
+        if ("true".contentEquals(state)) {
+            return OnOffType.ON;
+        } else if ("false".equals(state)) {
+            return OnOffType.OFF;
+        } else {
+            logger.error("Unexpected state: {}", state);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    public State queryUnitState(final String action, final String name, final String module, final Unit<?> unit) {
+        final String s = queryStringState(action, name, module);
+        try {
+            if (s == null) {
+                return UnDefType.UNDEF;
+            } else {
+                return new QuantityType<>(Double.valueOf(s), unit);
+            }
+        } catch (NumberFormatException e) {
+            logger.error("Unexpected number format: {}", s);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    public void setState(final String action, final String name, final String moduleID, final String nick,
+            final String description, final OnOffType state) {
+        setState(action, name, moduleID, nick, description, OnOffType.ON.equals(state) ? "true" : "false");
+    }
+
+    public void setState(final String action, final String name, final String moduleID, final String nick,
+            final String description, final String state) {
+        try {
+            final SOAPMessage message = newMessage(action, moduleParameters(moduleID),
+                    soapFactory.createElement("NickName").addTextNode(nick),
+                    soapFactory.createElement("Description").addTextNode(description),
+                    soapFactory.createElement("OPStatus").addTextNode(state));
+            final Document document = post(action, message, authentication);
+            final String response = getElementByTagName(document, name);
+            if (!"ok".equals(response.toLowerCase())) {
+                logger.error("Failed to set state.");
```

Comment:
Probably warn. You also might want to include some more context in the error message, like what state failed and with which value so the user has an idea where to look. The current message is useless as it gives no indication what is wrong.